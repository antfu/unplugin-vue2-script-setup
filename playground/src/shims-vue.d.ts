/// <reference types="unplugin-vue2-script-setup/shims.js" />
/// <reference types="unplugin-vue2-script-setup/ref-macros.js" />
