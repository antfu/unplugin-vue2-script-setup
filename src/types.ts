import { Program } from '@babel/types'

export interface ScriptTagMeta {
  start: number
  end: number
  contentStart: number
  contentEnd: number
  content: string
  attrs: Record<string, string>
  found: boolean
  ast: Program
}

export interface ParsedSFC {
  id?: string
  template: {
    components: Set<string>
    identifiers: Set<string>
  }
  scriptSetup: ScriptTagMeta
  script: ScriptTagMeta
}

export interface ScriptSetupTransformOptions {
  astTransforms?: {
    script?: (ast: Program) => Program
    scriptSetup?: (ast: Program) => Program
    post?: (ast: Program, sfc: ParsedSFC) => Program
  }
  refTransform?: boolean
}
