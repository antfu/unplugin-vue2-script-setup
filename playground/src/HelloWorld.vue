<script lang="ts">
/* eslint-disable import/first */
export default {
  name: 'App',
}
</script>
<script setup lang="ts">
import { watch } from '@vue/composition-api'
import Foo from './Foo.vue'
import Bar from './Bar.vue'

withDefaults(defineProps<{ msg: string; name: string | number }>(), { msg: 'Hello' })
const emit = defineEmits<{
  (event: 'update', value: number): void
}>()

let count = $ref(0)
// eslint-disable-next-line prefer-const
let doubled = $computed(() => count * 2)

function inc() {
  count += 1
}
function dec() {
  count -= 1
}

const decText = '<b>Dec</b>'

watch(count, value => emit('update', value))
</script>
<template>
  <div>
    <h3>{{ msg }}, {{ name }}</h3>
    <button @click="inc">
      Inc
    </button>
    <div>{{ count }} x 2 = {{ doubled }}</div>
    <button @click="dec()" v-html="decText" />
    <component :is="count > 2 ? Foo : Bar" />
  </div>
</template>
