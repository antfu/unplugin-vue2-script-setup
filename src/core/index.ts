export * from './transform'
export * from './transformReSugar'
export * from './transformScriptSetup'
export * from '../types'
