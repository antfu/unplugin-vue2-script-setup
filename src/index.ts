import { createUnplugin } from 'unplugin'
import { ScriptSetupTransformOptions } from './types'
import { transform } from '.'

export * from './core/transform'
export * from './types'

const scriptSetupRE = /<script\s(.*\s)?setup(\s.*)?>/

export default createUnplugin<ScriptSetupTransformOptions>(options => ({
  name: 'unplugin-vue2-script-setup',
  enforce: 'pre',
  transformInclude(id) {
    return id.endsWith('.vue')
  },
  transform(code, id) {
    try {
      if (scriptSetupRE.test(code))
        return transform(code, id, options)
    }
    catch (e) {
      this.error(e)
    }
  },
}))
