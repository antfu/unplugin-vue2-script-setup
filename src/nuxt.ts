import * as defu from 'defu'
import type { PluginOptions } from './types'
import unplugin from '.'

function scriptSetupModule(this: any, inlineOptions: PluginOptions = {}) {
  // FIXME: defu cjs types should changed
  const options = (defu.default || defu)(inlineOptions, this.nuxt.options.scriptSetup)

  // install webpack plugin
  this.extendBuild((config: any) => {
    config.plugins = config.plugins || []
    config.plugins.unshift(unplugin.webpack(options))
  })

  // install vite plugin
  this.nuxt.hook('vite:extend', async(vite: any) => {
    vite.config.plugins = vite.config.plugins || []
    vite.config.plugins.push(unplugin.vite(options))
  })
}

export default scriptSetupModule
