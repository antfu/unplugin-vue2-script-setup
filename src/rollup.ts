import unplugin from '.'

export default unplugin.rollup
module.exports = unplugin.rollup
module.exports.default = unplugin.rollup
Object.defineProperty(module.exports, '__esModule', { value: true })
