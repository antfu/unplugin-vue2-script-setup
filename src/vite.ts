import unplugin from '.'

export = unplugin.vite
