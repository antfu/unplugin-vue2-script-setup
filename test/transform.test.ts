import { resolve } from 'path'
import { promises as fs } from 'fs'
import fg from 'fast-glob'
import { transform } from '../src'

describe('transform', () => {
  describe('fixtures', () => {
    const root = resolve(__dirname, '..')
    const files = fg.sync([
      'test/fixtures/*.{vue,js,ts}',
      'playground/*.{vue,js,ts}',
    ], {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files) {
      it(file.replace(/\\/g, '/'), async() => {
        const warn = jest.spyOn(console, 'warn').mockImplementation(() => {})

        const fixture = await fs.readFile(resolve(root, file), 'utf-8')
        const result = transform(fixture, file, { refTransform: true })?.code || fixture
        expect(result).toMatchSnapshot()

        const result2 = transform(result, file, { refTransform: true })?.code || result
        expect(result).toEqual(result2)

        warn.mockRestore()
      })
    }
  })
})
