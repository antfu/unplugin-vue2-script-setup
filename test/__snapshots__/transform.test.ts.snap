// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform fixtures playground/App.vue 1`] = `
"<template>
  <div>
    <hello-world name=\\"Vue 2\\" @update=\\"onUpdate\\" />

    <async-component />
  </div>
</template>

<script lang=\\"ts\\">
import { defineAsyncComponent } from '@vue/composition-api';
import HelloWorld from './HelloWorld.vue';
const AsyncComponent = defineAsyncComponent(() => import('./Async.vue'));
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  function onUpdate(e: any) {
    // eslint-disable-next-line no-console
    console.log(e);
  }

  return {
    onUpdate
  };
};

__sfc_main.components = Object.assign({
  HelloWorld,
  AsyncComponent
}, __sfc_main.components);
export default __sfc_main;
</script>
"
`;

exports[`transform fixtures playground/Async.vue 1`] = `
"<template>
  <div>Async Component</div>
</template>
"
`;

exports[`transform fixtures playground/Bar.vue 1`] = `
"<template>
  <div>Bar</div>
</template>
"
`;

exports[`transform fixtures playground/Foo.vue 1`] = `
"<template>
  <div>Foo</div>
</template>
"
`;

exports[`transform fixtures playground/HelloWorld.vue 1`] = `
"<template>
  <div>
    <h3>{{ msg }}, {{ name }}</h3>
    <button @click=\\"inc\\">
      Inc
    </button>
    <div>{{ count }} x 2 = {{ doubled }}</div>
    <button @click=\\"dec()\\" v-html=\\"decText\\" />
    <component :is=\\"count > 2 ? Foo : Bar\\" />
  </div>
</template>

<script lang=\\"ts\\">
import { ref as _ref, computed as _computed } from \\"@vue/composition-api\\";
import { watch } from '@vue/composition-api';
import Foo from './Foo.vue';
import Bar from './Bar.vue';
const __sfc_main = {
  name: 'App'
};
__sfc_main.props = {
  msg: {
    key: \\"msg\\",
    required: false,
    type: String,
    default: 'Hello'
  },
  name: {
    key: \\"name\\",
    required: true,
    type: [String, Number]
  }
};

__sfc_main.setup = (__props, __ctx) => {
  const props = __props;
  const emit = __ctx.emit;

  let count = _ref(0); // eslint-disable-next-line prefer-const


  let doubled = _computed(() => count.value * 2);

  function inc() {
    count.value += 1;
  }

  function dec() {
    count.value -= 1;
  }

  const decText = '<b>Dec</b>';
  watch(count.value, value => emit('update', value));
  return {
    Foo,
    Bar,
    count,
    doubled,
    inc,
    dec,
    decText
  };
};

export default __sfc_main;
</script>


"
`;

exports[`transform fixtures playground/main.ts 1`] = `
"import Vue from 'vue'
import VueCompositionAPI from '@vue/composition-api'
import App from './App.vue'

Vue.use(VueCompositionAPI)

const app = new Vue({ render: h => h(App) })
app.$mount('#app')
"
`;

exports[`transform fixtures playground/vite.config.ts 1`] = `
"import { defineConfig } from 'vite'
import { createVuePlugin as Vue2 } from 'vite-plugin-vue2'
import Inspect from 'vite-plugin-inspect'
import ScriptSetup from '../src/vite'

export default defineConfig({
  plugins: [
    Vue2(),
    Inspect(),
    ScriptSetup({
      refTransform: true,
    }),
  ],
})
"
`;

exports[`transform fixtures test/fixtures/AsyncImport.vue 1`] = `
"<script lang=\\"ts\\">
import { defineAsyncComponent } from '@vue/composition-api';
const ScriptOnly = defineAsyncComponent(() => import('./ScriptOnly.vue'));
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  return {};
};

__sfc_main.components = Object.assign({
  ScriptOnly
}, __sfc_main.components);
export default __sfc_main;
</script>

<template>
  <ScriptOnly />
</template>
"
`;

exports[`transform fixtures test/fixtures/DynamicStyle.vue 1`] = `
"<template>
  <div :style=\\"{ color, border: '1px' }\\" />
</template>

<script lang=\\"ts\\">
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const color = computed(() => 'red');
  return {
    color
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/Empty.vue 1`] = `""`;

exports[`transform fixtures test/fixtures/Enum.vue 1`] = `
"<template>
  <div>{{ MyEnum }}</div>
</template>

<script lang=\\"ts\\">
enum MyEnum {
  test = 'true',
}
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  return {
    MyEnum
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/Macros.vue 1`] = `
"<template>
  <div @click=\\"emit(props.msg)\\">{{ msg }}</div>
</template>

<script lang=\\"js\\">
const __sfc_main = {};
__sfc_main.props = {
  msg: String
};

__sfc_main.setup = (__props, __ctx) => {
  const props = __props;
  const emit = __ctx.emit;
  return {
    props,
    emit
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/MacrosDefineExpose.vue 1`] = `
"<template>
  <div>{{ a }}</div>
  <div>{{ text?.length ?? textLengthDefault }}</div>
</template>

<script lang=\\"ts\\">
import { ref } from '@vue/composition-api';
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const a = ref(1);
  const b = 1;
  const text = 'hello';
  const textLengthDefault = 0;
  return Object.assign({
    a,
    text,
    textLengthDefault
  }, {
    b
  });
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/MacrosPure.vue 1`] = `
"<template>
  <div>{{ msg }}</div>
</template>

<script >
const __sfc_main = {};
__sfc_main.props = {
  msg: String
};
export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/MacrosType.vue 1`] = `
"<template>
  <div @click=\\"emit('update', props.msg)\\">
    {{ msg }}
  </div>
</template>

<script lang=\\"ts\\">
const __sfc_main = {};
__sfc_main.props = {
  msg: {
    key: \\"msg\\",
    required: false,
    type: String,
    default: 'Hello'
  },
  value: {
    key: \\"value\\",
    required: true,
    type: [Number, String]
  },
  data: {
    key: \\"data\\",
    required: false,
    type: Object
  }
};

__sfc_main.setup = (__props, __ctx) => {
  const props = __props;
  const emit = __ctx.emit;
  return {
    props,
    emit
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/MacrosType2.vue 1`] = `
"<template>
  <div @click=\\"emit(props.msg)\\">
    {{ msg }}
  </div>
</template>

<script lang=\\"ts\\">
export interface Props {
  msg: string;
  value: number | string;
  data?: {
    value: boolean;
  };
  arr?: [number, string, {}];
  any: any;
}
const __sfc_main = {};
__sfc_main.props = {
  msg: {
    key: \\"msg\\",
    required: false,
    type: String,
    default: 'Hello'
  },
  value: {
    key: \\"value\\",
    required: true,
    type: [Number, String]
  },
  data: {
    key: \\"data\\",
    required: false,
    type: Object
  },
  arr: {
    key: \\"arr\\",
    required: false,
    type: Array
  },
  any: {
    key: \\"any\\",
    required: true,
    type: null
  }
};

__sfc_main.setup = (__props, __ctx) => {
  const props = __props;
  const emit = __ctx.emit;
  return {
    props,
    emit
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/MacrosType3.vue 1`] = `
"<template>
  <div @click=\\"emit('update', msg)\\">
    {{ msg }}
  </div>
</template>

<script lang=\\"ts\\">
const __sfc_main = {};
__sfc_main.props = {
  msg: {
    key: \\"msg\\",
    required: false,
    type: String,
    default: 'Hello'
  },
  value: {
    key: \\"value\\",
    required: true,
    type: [Number, String]
  },
  data: {
    key: \\"data\\",
    required: false,
    type: Object
  }
};

__sfc_main.setup = (__props, __ctx) => {
  const emit = __ctx.emit;
  return {
    emit
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/MacrosType4.vue 1`] = `
"<script lang=\\"ts\\">
type Test = number[];
const __sfc_main = {};
__sfc_main.props = {
  test: {
    key: \\"test\\",
    required: false,
    type: null,
    default: () => []
  }
};

__sfc_main.setup = (__props, __ctx) => {
  return {};
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/MacrosTypeAny.vue 1`] = `
"<template>
  <div>
    {{ value }}
  </div>
</template>

<script lang=\\"ts\\">
const __sfc_main = {};
__sfc_main.props = {
  value: {
    key: \\"value\\",
    required: true,
    type: null
  }
};

__sfc_main.setup = (__props, __ctx) => {
  const props = __props;
  return {};
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/Object1.vue 1`] = `
"<template>
  <div>
    {{ data }} {{ title }}
  </div>
</template>

<script lang=\\"ts\\">
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const data = {
    title: 'Template Only'
  };
  return {
    data
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/ObjectDestructure.vue 1`] = `
"<template>
  <div></div>
</template>

<script >
import { toRefs, reactive } from 'vue';
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const state = reactive({
    data: 'hello'
  });
  const {
    data
  } = toRefs(state);
  return {};
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/Pug1.vue 1`] = `
"<template lang=\\"pug\\">
.root
  span {{ data }} {{ title }}
  .p-1(@click=\\"foo\\")
</template>

<script lang=\\"ts\\">
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const data = {
    title: 'Template Only'
  };

  function foo() {}

  return {
    data,
    foo
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/RefSugar.ts 1`] = `
"import { ref as _ref, computed as _computed } from '@vue/composition-api'
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable prefer-const */
import { ref } from '@vue/composition-api'

let ref1 = (ref('hello'))
ref1.value = 'world'
let ref1Raw = (ref1)

let ref2 = _ref('hello')
ref2.value = 'world'
let ref2Raw = (ref2)

let computed1 = _computed(() => ref2.value += ' str')
let computed1Raw = (computed1)
"
`;

exports[`transform fixtures test/fixtures/RefSugar.vue 1`] = `
"<template>
  <div>
    {{ msg }}
    {{ msg2 }}
    {{ a }}
  </div>
</template>



<script lang=\\"ts\\">
import { ref as _ref } from \\"@vue/composition-api\\";

/* eslint-disable prefer-const */
import { defineComponent, ref } from '@vue/composition-api';

const __sfc_main = defineComponent({
  setup() {
    let msg = ref('hello world');
    return {
      msg: msg.value,
      msgRef: msg
    };
  }

});

__sfc_main.setup = (__props, __ctx) => {
  let msg2 = _ref('hello world');

  let a: number;
  a = 2;
  return {
    msg2,
    a
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/RefSugarScriptSetup.vue 1`] = `
"<template>
  <div>
    {{ msg }}
  </div>
</template>

<script lang=\\"ts\\">
import { ref as _ref } from \\"@vue/composition-api\\";
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  /* eslint-disable prefer-const */
  let msg = _ref('hello world');

  return {
    msg
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/ScriptLessThanOrEqualTo.vue 1`] = `
"<script lang=\\"ts\\">
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const val: boolean = 1 <= 2;
  return {
    val
  };
};

export default __sfc_main;
</script>

<template>
  <div>{{ val }}</div>
</template>
"
`;

exports[`transform fixtures test/fixtures/ScriptOnly.vue 1`] = `
"<script lang=\\"ts\\">
// <script setup> in comment should not be matched
import { defineComponent } from '@vue/composition-api';

const __sfc_main = defineComponent({
  name: 'Hi'
});

export default __sfc_main;
</script>


<!-- <script setup> -->
"
`;

exports[`transform fixtures test/fixtures/TemplateOnly.vue 1`] = `
"<template>
  <div>
    <div v-for=\\"(item, index) in items\\" :key=\\"item\\">
      {{ index }}: {{ item }}
    </div>
  </div>
</template>

<script lang=\\"ts\\">
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const items = [{
    name: 'Item 1',
    value: 1
  }, {
    name: 'Item 2',
    value: 2
  }, {
    name: 'Item 3',
    value: 3
  }, {
    name: 'Item 4',
    value: 4
  }];
  return {
    items
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/TemplateOptionalChaining.vue 1`] = `
"<template>
  <div @click=\\"callback?.()\\">
    {{ text?.length ?? textLengthDefault }}
    {{ classes?.[0] }}
  </div>
</template>

<script lang=\\"ts\\">
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const text = 'hello';
  const textLengthDefault = 0;
  const callback = (undefined as undefined | (() => void));
  const classes = (undefined as undefined | string[]);
  return {
    text,
    textLengthDefault,
    callback,
    classes
  };
};

export default __sfc_main;
</script>
"
`;

exports[`transform fixtures test/fixtures/VFor.vue 1`] = `
"<template>
  <div>
    <div v-for=\\"(item, index) in items\\" :key=\\"item\\">
      {{ item }}
    </div>
    <div v-for=\\"{ cargo_key, in_of } in items2\\">
      {{ cargo_key }}
    </div>
  </div>
</template>

<script lang=\\"ts\\">
const __sfc_main = {};

__sfc_main.setup = (__props, __ctx) => {
  const items = [{
    name: 'Item 1',
    value: 1
  }, {
    name: 'Item 2',
    value: 2
  }, {
    name: 'Item 3',
    value: 3
  }, {
    name: 'Item 4',
    value: 4
  }];
  const items2 = [{
    cargo_key: 'cargo_key',
    in_of: 'in'
  }];
  const index = 0;
  return {
    items,
    items2
  };
};

export default __sfc_main;
</script>
"
`;
